snap/naming/tag.go

```diff
@@ -0,0 +1,125 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package naming
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+var errInvalidSecurityTag = errors.New("invalid security tag")
+
+// ParsedSecurityTag describes a parsed snap security tag.
+type ParsedSecurityTag interface {
+	// String returns the entire security tag.
+	String() string
+
+	// InstanceName returns the snap name and instance key.
+	InstanceName() string
+}
+
+// ParsedAppSecurityTag describes a parsed snap application security tag.
+type ParsedAppSecurityTag interface {
+	ParsedSecurityTag
+	// AppName returns the name of the application.
+	AppName() string
+}
+
+type appSecurityTag struct {
+	instanceName string
+	appName      string
+}
+
+func (t appSecurityTag) String() string {
+	return fmt.Sprintf("snap.%s.%s", t.instanceName, t.appName)
+}
+
+func (t appSecurityTag) InstanceName() string {
+	return t.instanceName
+}
+
+func (t appSecurityTag) AppName() string {
+	return t.appName
+}
+
+// ParsedAppSecurityTag describes a parsed snap hook security tag.
+type ParsedHookSecurityTag interface {
+	ParsedSecurityTag
+	// HookName returns the name of the hook.
+	HookName() string
+}
+
+type hookSecurityTag struct {
+	instanceName string
+	hookName     string
+}
+
+func (t hookSecurityTag) String() string {
+	return fmt.Sprintf("snap.%s.hook.%s", t.instanceName, t.hookName)
+}
+
+func (t hookSecurityTag) InstanceName() string {
+	return t.instanceName
+}
+
+func (t hookSecurityTag) HookName() string {
+	return t.hookName
+}
+
+// ParseSecurityTag parses a snap security tag and returns a parsed representation or an error.
+//
+// Further type assertions can be used to described the particular form, either
+// describing an application or a hook specific security tag.
+func ParseSecurityTag(tag string) (ParsedSecurityTag, error) {
+	// We expect at most four parts. Split with up to five parts so that the
+	// len(parts) test catches invalid format tags very early.
+	parts := strings.SplitN(tag, ".", 5)
+	// We expect either three or four components.
+	if len(parts) != 3 && len(parts) != 4 {
+		return nil, errInvalidSecurityTag
+	}
+	// We expect "snap" and the snap instance name as first two fields.
+	snapLiteral, snapName := parts[0], parts[1]
+	if snapLiteral != "snap" {
+		return nil, errInvalidSecurityTag
+	}
+	if err := ValidateInstance(snapName); err != nil {
+		return nil, errInvalidSecurityTag
+	}
+	// Depending on the type of the tag we either expect application name or
+	// the "hook" literal and the hook name.
+	if len(parts) == 3 {
+		appName := parts[2]
+		if err := ValidateApp(appName); err != nil {
+			return nil, errInvalidSecurityTag
+		}
+		return &appSecurityTag{instanceName: snapName, appName: appName}, nil
+	} else {
```

Comment:
I think govet complains about having this indented, it says to de-dent this block and drop the else since the if branch ends with a return.

Reply:
This is a conversion from a switch statement, that drops an unused default case. I think what you suggest is fine, for as long as we don't have more cases to add. I know there's one weird security tag that we have, that is used in one specific spot, that is not covered by this code yet but I will deal with it when it comes to that.